vioscsi/helper.c

```diff
@@ -111,7 +112,9 @@ ENTER_FN_SRB();
         }
```

Comment:
Thinking of what could get us in this situation, this would happen if we've put a bunch of IO on the ring, filled it up, and the host virtio thread(s) have not reaped it fast enough.

Assuming a split queue setup -> virtqueue_add_buf_split will return only return res < 0 like so:

```
    if (va_indirect && (out + in) > 1 && vq->num_unused > 0) { <<< must be false
        ...
    } else {
        u16 last_idx;

        /* Use out + in regular descriptors */
        if (out + in > vq->num_unused) { <<<< must be true
            return -ENOSPC; <<<<<<<<<
        }
        ...
```

In the successful path, however, we're always calling get_unused_desc(), which decrements num_unused.

This means that we could fill the queue up, and if the host has kick suppression turned on, it won't get a notice that the ring is completely full until the //next// IO comes in. 

In reality, we should (probably) kick when the VQ is 100% full (or perhaps even num_unused <= 1), not 100% + 1 entry full, such that we could potentially (sometimes) avoid this situation in the first place?

i.e. in virtqueue_kick_prepare_split():
  return !(vq->vring.used->flags & VIRTQ_USED_F_NO_NOTIFY) || vq->num_unused <= 1;

We could take that patch separately as an optimization to this one, as I think this patch makes 100% sense based on the API documentation you linked

Reply:
@vrozenfe - thoughts on the additional logic for kicks noted above?

Reply:
@JonKohler 
Hi Jon, I fully agree that we should try to minimize the chances of getting into situation when a VQ is (over)saturated.
Ideally, it would be preferable if we can rely only on the mechanisms provided by Storport (like StorPortSetDeviceQueueDepth , MaxIOsPerLun, etc).  But since it doesn't work for all cases and we still can hit the case when VQ is filled up, your idea to introduce some sort of threshold to notify VQ before that looks very reasonable to me.

All the best,
Vadim.